### PR TITLE
BUG: contour plot levels

### PR DESCRIPTION
**Pull request by fmaussion**
This should fix https://github.com/pydata/xarray/issues/866

### CODE REVIEW

File: xarray/plot/plot.py [hunk 1]
```diff
@@ -387,6 +387,18 @@ def newplotfunc(darray, x=None, y=None, ax=None, row=None, col=None,
...
+        if plotfunc.__name__ == 'contour':
+            # Contouring data is equivalent to extend upwards
+            extend = extend or 'max'
+
+            # A colorbar with one level is not possible with mpl
+            try:
+                if len(levels) < 2:
+                    add_colorbar = False
+            except TypeError:
+                if levels < 2:
```
Comment (fmaussion) [inline xarray/plot/plot.py]: This is a huge corner case, but well...

File: xarray/plot/plot.py [hunk 2]
```diff
@@ -384,8 +384,16 @@ def newplotfunc(darray, x=None, y=None, ax=None, row=None, col=None,
 
         _ensure_plottable(xval, yval)
 
-        if 'contour' in plotfunc.__name__ and levels is None:
-            levels = 7  # this is the matplotlib default
+        if 'contour' in plotfunc.__name__:
+            if levels is None:
+                levels = 7  # this is the matplotlib default
+            # A colorbar with one level is not possible with mpl
+            try:
```
Comment (shoyer) [inline xarray/plot/plot.py]: Can we do a type check instead? e.g.,
```
if (isinstance(levels, int) and levels < 2) or len(levels) < 2:
    add_colorbar = False
```

Comment (shoyer) [inline xarray/plot/plot.py]: We could also change the default to `add_colorbar=None`, and only apply this logic if `add_colorbar` wasn't set directly.

**Comment by shoyer** [comment]
I wonder -- would it maybe make more sense to simply *not* to add a colorbar by default? It's pretty easy to toggle if desired.

**Comment by fmaussion** [comment]
@shoyer yes, this was my original idea too: `contour` plots usually don't need a colorbar. I changed the default to false for `contour` plots only, and removed the convoluted logic for when the user gives only one level: this is a matplotlib problem after all, and if the user sets `levels=1` and `add_colorbar=True` it really isn't xarray's problem anymore.

I think it's better now, let me know what you think.

File: xarray/plot/utils.py [hunk 3]
```diff
@@ -41,7 +41,7 @@ def _build_discrete_cmap(cmap, levels, extend, filled):
 
     if not filled:
         # non-filled contour plots
-        extend = 'neither'
+        extend = 'max'
```
Comment (shoyer) [inline xarray/plot/utils.py]: I think I missed the rationale for this one -- why should this change?

Comment (fmaussion) [inline xarray/plot/utils.py]: This is the actual bug fix for https://github.com/pydata/xarray/issues/866 which motivated this PR ;-).

xarray wasn't plotting contours correctly before this change, see the following code:

```python
import numpy as np
import matplotlib.pyplot as plt
import xarray as xr

x, y = np.meshgrid(np.arange(12), np.arange(12))
z = xr.DataArray(np.sqrt(x**2 + y**2))
z.plot.contour(levels=[2, 4, 6, 8], add_colorbar=False, colors='k')
plt.show()
```
figure_1

With this fix everything is fine now:

figure_1-1

Intuitively, it also makes sense: the last contour does mark all data beyond the last level, hence `extend = 'max'`

Comment (shoyer) [inline xarray/plot/utils.py]: Got it, thanks!

**Pull request merged**